specs/index.md

```diff
@@ -0,0 +1,312 @@
+
+## Design Principles
+
+The design of the indexing system considers the following key aspects:
+
+- **Upgradability & Versioning:**
+  - Index schemas are versioned to allow future enhancements while maintaining compatibility.
+  - Future optimizations might include truncating stored multihashes to the last six characters, as content can be validated by re-hashing.
+  - Indexes could store only byte ranges in the DAG rather than intermediary multihashes, requiring clients to provide context at request time on where they are at.
+- **Modular & Pluggable Indexing:**
+  - The protocol supports multiple indexing strategies based on different trade-offs.
+  - Implementations can choose the best index type depending on cost, performance, and use case.
+  - Supports integrating new indexing mechanisms without requiring changes to the core protocol.
+- **Storage Efficiency & Cost Optimization:**
+  - The protocol aims to minimize storage and retrieval costs by leveraging compact and structured index formats.
+  - Different backends (e.g., file-based, JSON, SQL, external CARv2 indexes) provide flexible storage options.
+
+## Index Types
+
+### Block-level index
+
+A Block-level index maps a given block multihash to the location where the server can read the bytes of the block. This indexing strategy enables fast responses for clients requesting verifiable blocks by a block CID. However, in some setups, implementing this indexing strategy may be prohibitively expensive (e.g., in databases like DynamoDB) or limited (e.g., due to rate limiting when indexing thousands of blocks in parallel from large containers). Furthermore, these indexes alone are insufficient to serve fully verifiable containers, as they do not maintain relationships between blocks in a DAG.
+
+#### Block-level Type index Schema
+
+```ts
+type Index = Variant<{
+  'index/block@0.1': BlockIndex
+}>
+
+type BlockIndex = {
+  // hash digest of the container containing the block
+  digest: Multihash
+  // Slice offset
+  offset: Int
+  // Slice size in bytes
+  length: Int
+}
+
+type Multihash = bytes
+```
+
+### Multiple-level index
+
+A Multiple-level index maps a content multihash to a list of verifiable package CIDs containing the blocks that form the DAG for that content (and where they are positioned). This approach allows serving fully verifiable containers efficiently while reducing index store operations by several orders of magnitude. However, this index alone cannot serve block-level requests unless the request includes hints about the content CID context.
+
+#### Multiple-level Type index Schema
+
+```ts
+type Index = Variant<{
+  "index/sharded/dag@0.1": ShardedDAGIndex
+}>
+
+type ShardedDAGIndex = {
+  // content root CID
+  content: Link<any>
```

Comment:
would be great to be multihash to be honest, but then would not be compatible with w3s. thoughts?

Reply:
At this point, I am questioning how much compatibility we would like here.

If we would use multihash, Storacha index could upgrade to this one, but not other way around. If we want the Index N:M at this level, content may make sense though, open for feedback.